File: ios/VLCPro/VLCPlayerView.swift
import Foundation
import UIKit
import React
import MobileVLCKit

/**
 * Vue native VLCPlayer pour iOS
 * Phase 1, Semaine 2 : Intégration MobileVLCKit complète
 */
@objc(VLCPlayerView)
class VLCPlayerView: RCTView {
    
    // MARK: - Properties
    
    private var mediaPlayer: VLCMediaPlayer?
    private var vlcLibrary: VLCLibrary?
    private var currentMedia: VLCMedia?
    
    // React Native props
    @objc var source: NSDictionary? {
        didSet {
            if let source = source {
                loadMedia(from: source)
            }
        }
    }
    
    @objc var autoPlay: Bool = false {
        didSet {
            if autoPlay && mediaPlayer?.media != nil {
                play()
            }
        }
    }
    
    @objc var loop: Bool = false
    @objc var muted: Bool = false {
        didSet {
            mediaPlayer?.audio?.isMuted = muted
        }
    }
    
    @objc var volume: NSNumber = 100 {
        didSet {
            let volumeValue = Int32(volume.intValue)
            mediaPlayer?.audio?.volume = volumeValue
        }
    }
    
    @objc var rate: NSNumber = 1.0 {
        didSet {
            mediaPlayer?.rate = rate.floatValue
        }
    }
    
    @objc var paused: Bool = false {
        didSet {
            if paused {
                pause()
            } else {
                play()
            }
        }
    }
    
    // Event callbacks
    @objc var onReady: RCTDirectEventBlock?
    @objc var onPlay: RCTDirectEventBlock?
    @objc var onPause: RCTDirectEventBlock?
    @objc var onStop: RCTDirectEventBlock?
    @objc var onEnd: RCTDirectEventBlock?
    @objc var onError: RCTDirectEventBlock?
    @objc var onProgress: RCTDirectEventBlock?
    @objc var onLoad: RCTDirectEventBlock?
    @objc var onBuffer: RCTDirectEventBlock?
    @objc var onStateChange: RCTDirectEventBlock?
    @objc var onGetCurrentTime: RCTDirectEventBlock?
    @objc var onGetDuration: RCTDirectEventBlock?
    @objc var onGetState: RCTDirectEventBlock?
    @objc var onGetVolume: RCTDirectEventBlock?
    @objc var onIsMuted: RCTDirectEventBlock?
    @objc var onGetRate: RCTDirectEventBlock?
    
    // MARK: - Lifecycle
    
    override init(frame: CGRect) {
        super.init(frame: frame)
        setupVLC()
    }
    
    required init?(coder: NSCoder) {
        super.init(coder: coder)
        setupVLC()
    }
    
    deinit {
        cleanup()
    }
    
    // MARK: - VLC Setup
    
    private func setupVLC() {
        // Initialiser VLCLibrary
        vlcLibrary = VLCLibrary.shared()
        
        // Créer le media player
        mediaPlayer = VLCMediaPlayer(library: vlcLibrary)
        
        // Configurer la vue de sortie
        if let mediaPlayer = mediaPlayer {
            mediaPlayer.drawable = self
            mediaPlayer.delegate = self
            
            // Configuration par défaut
            mediaPlayer.audio?.isMuted = muted
            mediaPlayer.audio?.volume = Int32(volume.intValue)
            mediaPlayer.rate = rate.floatValue
        }
        
        // Configuration de la vue
        backgroundColor = UIColor.black
        clipsToBounds = true
    }
    
    // MARK: - Media Loading
    
    private func loadMedia(from source: NSDictionary) {
        guard let uri = source["uri"] as? String else {
            sendError(message: "URI manquante dans la source", code: "INVALID_SOURCE")
            return
        }
        
        var media: VLCMedia?
        
        if uri.hasPrefix("http://") || uri.hasPrefix("https://") {
            // URL réseau
            if let url = URL(string: uri) {
                media = VLCMedia(url: url)
                
                // Ajouter les headers si présents
                if let headers = source["headers"] as? [String: String] {
                    for (key, value) in headers {
                        media?.addOption("\(key)=\(value)")
                    }
                }
            }
        } else if uri.hasPrefix("file://") {
            // Fichier local
            if let url = URL(string: uri) {
                media = VLCMedia(url: url)
            }
        } else {
            // Chemin relatif dans le bundle
            if let path = Bundle.main.path(forResource: uri, ofType: nil) {
                media = VLCMedia(path: path)
            }
        }
        
        guard let validMedia = media else {
            sendError(message: "Impossible de créer le média à partir de l'URI: \(uri)", code: "MEDIA_CREATION_FAILED")
            return
        }
        
        currentMedia = validMedia
        mediaPlayer?.media = validMedia
        
        // Démarrer la lecture automatique si activée
        if autoPlay {
            play()
        }
    }
    
    // MARK: - Playback Controls
    
    @objc func play() {
        guard let mediaPlayer = mediaPlayer else { return }
        
        if mediaPlayer.media == nil {
            sendError(message: "Aucun média chargé", code: "NO_MEDIA")
            return
        }
        
        mediaPlayer.play()
    }
    
    @objc func pause() {
        mediaPlayer?.pause()
    }
    
    @objc func stop() {
        mediaPlayer?.stop()
    }
    
    @objc func seek(to time: NSNumber) {
        guard let mediaPlayer = mediaPlayer else { return }
        
        let timeMs = time.doubleValue
        let timeSeconds = timeMs / 1000.0
        let duration = mediaPlayer.media?.length.doubleValue ?? 0
        
        if duration > 0 {
            let position = Float(timeSeconds / (duration / 1000.0))
            mediaPlayer.position = min(max(position, 0.0), 1.0)
        }
    }
    
    // MARK: - Information Methods
    
    @objc func getCurrentTime() {
        guard let mediaPlayer = mediaPlayer else { return }
        
        let currentTime = mediaPlayer.time.doubleValue
        let data: [String: Any] = ["currentTime": currentTime]
        sendEvent("onGetCurrentTime", data: data)
    }
    
    @objc func getDuration() {
        guard let mediaPlayer = mediaPlayer else { return }
        
        let duration = mediaPlayer.media?.length.doubleValue ?? 0
        let data: [String: Any] = ["duration": duration]
        sendEvent("onGetDuration", data: data)
    }
    
    @objc func getState() {
        let state = getCurrentState()
        let data: [String: Any] = ["state": state]
        sendEvent("onGetState", data: data)
    }
    
    @objc func getVolume() {
        guard let mediaPlayer = mediaPlayer else { return }
        
        let volume = mediaPlayer.audio?.volume ?? 100
        let data: [String: Any] = ["volume": volume]
        sendEvent("onGetVolume", data: data)
    }
    
    @objc func isMuted() {
        guard let mediaPlayer = mediaPlayer else { return }
        
        let muted = mediaPlayer.audio?.isMuted ?? false
        let data: [String: Any] = ["muted": muted]
        sendEvent("onIsMuted", data: data)
    }
    
    @objc func getRate() {
        guard let mediaPlayer = mediaPlayer else { return }
        
        let rate = mediaPlayer.rate
        let data: [String: Any] = ["rate": rate]
        sendEvent("onGetRate", data: data)
    }
    
    // MARK: - State Management
    
    private func getCurrentState() -> String {
        guard let mediaPlayer = mediaPlayer else { return "idle" }
        
        switch mediaPlayer.state {
        case .stopped:
            return "stopped"
        case .opening:
            return "loading"
        case .buffering:
            return "buffering"
        case .playing:
            return "playing"
        case .paused:
            return "paused"
        case .ended:
            return "ended"
        case .error:
            return "error"
        default:
            return "idle"
        }
    }
    
    // MARK: - Event Handling
    
    private func sendEvent(_ eventName: String, data: [String: Any] = [:]) {
        var eventData = data
        eventData["target"] = reactTag
        
        switch eventName {
        case "onReady":
            onReady?(eventData)
        case "onPlay":
            onPlay?(eventData)
        case "onPause":
            onPause?(eventData)
        case "onStop":
            onStop?(eventData)
        case "onEnd":
            onEnd?(eventData)
        case "onError":
            onError?(eventData)
        case "onProgress":
            onProgress?(eventData)
        case "onLoad":
            onLoad?(eventData)
        case "onBuffer":
            onBuffer?(eventData)
        case "onStateChange":
            onStateChange?(eventData)
        case "onGetCurrentTime":
            onGetCurrentTime?(eventData)
        case "onGetDuration":
            onGetDuration?(eventData)
        case "onGetState":
            onGetState?(eventData)
        case "onGetVolume":
            onGetVolume?(eventData)
        case "onIsMuted":
            onIsMuted?(eventData)
        case "onGetRate":
            onGetRate?(eventData)
        default:
            break
        }
    }
    
    private func sendError(message: String, code: String) {
        let errorData: [String: Any] = [
            "error": [
                "code": code,
                "message": message,
                "domain": "VLCPlayer"
            ]
        ]
        sendEvent("onError", data: errorData)
    }
    
    // MARK: - Cleanup
    
    private func cleanup() {
        mediaPlayer?.stop()
        mediaPlayer?.delegate = nil
        mediaPlayer = nil
        currentMedia = nil
    }
}

// MARK: - VLCMediaPlayerDelegate

extension VLCPlayerView: VLCMediaPlayerDelegate {
    
    func mediaPlayerStateChanged(_ aNotification: Notification) {
        DispatchQueue.main.async { [weak self] in
            guard let self = self else { return }
            
            let state = self.getCurrentState()
            let stateData: [String: Any] = [
                "state": state,
                "previousState": "unknown" // TODO: Track previous state
            ]
            
            self.sendEvent("onStateChange", data: stateData)
            
            // Envoyer des événements spécifiques
            switch state {
            case "playing":
                self.sendEvent("onPlay")
            case "paused":
                self.sendEvent("onPause")
            case "stopped":
                self.sendEvent("onStop")
            case "ended":
                if self.loop {
                    self.play()
                } else {
                    self.sendEvent("onEnd")
                }
            case "error":
                self.sendError(message: "Erreur de lecture VLC", code: "PLAYBACK_ERROR")
            default:
                break
            }
        }
    }
    
    func mediaPlayerTimeChanged(_ aNotification: Notification) {
        DispatchQueue.main.async { [weak self] in
            guard let self = self,
                  let mediaPlayer = self.mediaPlayer else { return }
            
            let currentTime = mediaPlayer.time.doubleValue
            let duration = mediaPlayer.media?.length.doubleValue ?? 0
            let position = mediaPlayer.position
            
            let progressData: [String: Any] = [
                "currentTime": currentTime,
                "duration": duration,
                "position": Double(position),
                "remainingTime": max(0, duration - currentTime)
            ]
            
            self.sendEvent("onProgress", data: progressData)
        }
    }
    
    func mediaPlayerMediaChanged(_ aNotification: Notification) {
        DispatchQueue.main.async { [weak self] in
            guard let self = self,
                  let mediaPlayer = self.mediaPlayer,
                  let media = mediaPlayer.media else { return }
            
            let loadData: [String: Any] = [
                "duration": media.length.doubleValue,
                "canPlayFast": true,
                "canPlaySlow": true,
                "canPause": true,
                "canSeek": true
            ]
            
            self.sendEvent("onLoad", data: loadData)
            self.sendEvent("onReady")
        }
    }
} 

File: android/src/main/java/com/vlcpro/VLCPlayerView.java
package com.vlcpro;

import android.app.Activity;
import android.app.PictureInPictureParams;
import android.content.Context;
import android.net.Uri;
import android.os.Build;
import android.util.AttributeSet;
import android.util.Rational;
import android.view.SurfaceHolder;
import android.view.SurfaceView;
import android.view.View;
import android.view.ViewGroup;
import android.view.Window;
import android.view.WindowManager;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.facebook.react.bridge.Arguments;
import com.facebook.react.bridge.ReactContext;
import com.facebook.react.bridge.ReadableMap;
import com.facebook.react.bridge.ReadableMapKeySetIterator;
import com.facebook.react.bridge.WritableMap;
import com.facebook.react.uimanager.events.RCTEventEmitter;

import org.videolan.libvlc.LibVLC;
import org.videolan.libvlc.Media;
import org.videolan.libvlc.MediaPlayer;
import org.videolan.libvlc.util.VLCVideoLayout;

import java.util.ArrayList;

/**
 * Vue native VLCPlayer pour Android
 * Phase 1, Semaine 2 : Intégration libVLC complète
 */
public class VLCPlayerView extends ViewGroup implements MediaPlayer.EventListener {
    
    // VLC Components
    private LibVLC libVLC;
    private MediaPlayer mediaPlayer;
    private VLCVideoLayout videoLayout;
    private SurfaceView surfaceView;
    
    // React Native props
    private ReadableMap source;
    private boolean autoPlay = false;
    private boolean loop = false;
    private boolean muted = false;
    private int volume = 100;
    private float rate = 1.0f;
    private boolean paused = false;
    
    // State
    private String currentState = "idle";
    private long currentTime = 0;
    private long duration = 0;
    private boolean isFullscreen = false;
    
    public VLCPlayerView(Context context) {
        super(context);
        init();
    }
    
    public VLCPlayerView(Context context, AttributeSet attrs) {
        super(context, attrs);
        init();
    }
    
    private void init() {
        setupVLC();
        setupVideoLayout();
    }
    
    private void setupVLC() {
        try {
            // Initialiser libVLC
            ArrayList<String> options = new ArrayList<>();
            options.add("--aout=opensles");
            options.add("--audio-time-stretch");
            options.add("-vvv");
            
            libVLC = new LibVLC(getContext(), options);
            
            // Créer le media player
            mediaPlayer = new MediaPlayer(libVLC);
            mediaPlayer.setEventListener(this);
            
        } catch (Exception e) {
            sendError("VLC_INIT_ERROR", "Erreur lors de l'initialisation de VLC: " + e.getMessage());
        }
    }
    
    private void setupVideoLayout() {
        // Créer le layout vidéo
        videoLayout = new VLCVideoLayout(getContext());
        
        // Créer la surface view
        surfaceView = new SurfaceView(getContext());
        videoLayout.addView(surfaceView);
        
        // Ajouter à la vue
        addView(videoLayout);
        
        // Attacher le media player à la surface
        if (mediaPlayer != null) {
            mediaPlayer.attachViews(videoLayout, null, false, false);
        }
    }
    
    @Override
    protected void onLayout(boolean changed, int l, int t, int r, int b) {
        if (videoLayout != null) {
            videoLayout.layout(0, 0, r - l, b - t);
        }
    }
    
    @Override
    protected void onMeasure(int widthMeasureSpec, int heightMeasureSpec) {
        super.onMeasure(widthMeasureSpec, heightMeasureSpec);
        if (videoLayout != null) {
            videoLayout.measure(widthMeasureSpec, heightMeasureSpec);
        }
    }
    
    // MARK: - Props Setters
    
    public void setSource(ReadableMap source) {
        this.source = source;
        loadMedia();
    }
    
    public void setAutoPlay(boolean autoPlay) {
        this.autoPlay = autoPlay;
        if (autoPlay && mediaPlayer != null && mediaPlayer.getMedia() != null) {
            play();
        }
    }
    
    public void setLoop(boolean loop) {
        this.loop = loop;
    }
    
    public void setMuted(boolean muted) {
        this.muted = muted;
        if (mediaPlayer != null) {
            mediaPlayer.setVolume(muted ? 0 : volume);
        }
    }
    
    public void setVolume(int volume) {
        this.volume = volume;
        if (mediaPlayer != null && !muted) {
            mediaPlayer.setVolume(volume);
        }
    }
    
    public void setRate(float rate) {
        this.rate = rate;
        if (mediaPlayer != null) {
            mediaPlayer.setRate(rate);
        }
    }
    
    public void setPaused(boolean paused) {
        this.paused = paused;
        if (paused) {
            pause();
        } else {
            play();
        }
    }
    
    // MARK: - Media Loading
    
    private void loadMedia() {
        if (source == null || !source.hasKey("uri")) {
            sendError("INVALID_SOURCE", "URI manquante dans la source");
            return;
        }
        
        String uri = source.getString("uri");
        if (uri == null || uri.isEmpty()) {
            sendError("INVALID_SOURCE", "URI vide");
            return;
        }
        
        try {
            Media media;
            
            if (uri.startsWith("http://") || uri.startsWith("https://")) {
                // URL réseau
                media = new Media(libVLC, Uri.parse(uri));
            } else if (uri.startsWith("file://")) {
                // Fichier local
                media = new Media(libVLC, Uri.parse(uri));
            } else {
                // Asset ou chemin relatif
                String assetPath = "android_asset://" + uri;
                media = new Media(libVLC, Uri.parse(assetPath));
            }
            
            // Ajouter les options si présentes
            if (source.hasKey("headers")) {
                ReadableMap headers = source.getMap("headers");
                if (headers != null) {
                    // Support des headers HTTP pour les streams réseau
                    ReadableMapKeySetIterator iterator = headers.keySetIterator();
                    while (iterator.hasNextKey()) {
                        String key = iterator.nextKey();
                        String value = headers.getString(key);
                        if (value != null) {
                            // Mapper les headers selon les options VLC
                            if (key.equalsIgnoreCase("User-Agent")) {
                                media.addOption(":http-user-agent=" + value);
                            } else if (key.equalsIgnoreCase("Authorization")) {
                                media.addOption(":http-password=" + value);
                            } else if (key.equalsIgnoreCase("Referer")) {
                                media.addOption(":http-referrer=" + value);
                            } else if (key.equalsIgnoreCase("Cookie")) {
                                media.addOption(":http-cookies=" + value);
                            } else {
                                // Headers personnalisés via http-forward-cookies
                                media.addOption(":http-forward-cookies");
                                media.addOption(":http-" + key.toLowerCase().replace("-", "") + "=" + value);
                            }
                        }
                    }
                }
            }
            
            mediaPlayer.setMedia(media);
            media.release();
            
            // Démarrer la lecture automatique si activée
            if (autoPlay) {
                play();
            }
            
        } catch (Exception e) {
            sendError("MEDIA_CREATION_FAILED", "Impossible de créer le média: " + e.getMessage());
        }
    }
    
    // MARK: - Playback Controls
    
    public void play() {
        if (mediaPlayer != null) {
            if (mediaPlayer.getMedia() == null) {
                sendError("NO_MEDIA", "Aucun média chargé");
                return;
            }
            mediaPlayer.play();
        }
    }
    
    public void pause() {
        if (mediaPlayer != null) {
            mediaPlayer.pause();
        }
    }
    
    public void stop() {
        if (mediaPlayer != null) {
            mediaPlayer.stop();
        }
    }
    
    public void seek(long timeMs) {
        if (mediaPlayer != null && duration > 0) {
            float position = (float) timeMs / duration;
            mediaPlayer.setPosition(Math.max(0.0f, Math.min(1.0f, position)));
        }
    }
    
    // MARK: - Information Methods
    
    public void getCurrentTime() {
        if (mediaPlayer != null) {
            currentTime = mediaPlayer.getTime();
            WritableMap data = Arguments.createMap();
            data.putDouble("currentTime", currentTime);
            sendEvent("onGetCurrentTime", data);
        }
    }
    
    public void getDuration() {
        if (mediaPlayer != null) {
            duration = mediaPlayer.getLength();
            WritableMap data = Arguments.createMap();
            data.putDouble("duration", duration);
            sendEvent("onGetDuration", data);
        }
    }
    
    public void getState() {
        WritableMap data = Arguments.createMap();
        data.putString("state", currentState);
        sendEvent("onGetState", data);
    }
    
    public void getVolume() {
        WritableMap data = Arguments.createMap();
        data.putInt("volume", volume);
        sendEvent("onGetVolume", data);
    }
    
    public void isMuted() {
        WritableMap data = Arguments.createMap();
        data.putBoolean("muted", muted);
        sendEvent("onIsMuted", data);
    }
    
    public void getRate() {
        WritableMap data = Arguments.createMap();
        data.putDouble("rate", rate);
        sendEvent("onGetRate", data);
    }
    
    // MARK: - State Management
    
    private String getStateFromVLC(int vlcState) {
        switch (vlcState) {
            case MediaPlayer.Event.Stopped:
                return "stopped";
            case MediaPlayer.Event.Opening:
                return "loading";
            case MediaPlayer.Event.Buffering:
                return "buffering";
            case MediaPlayer.Event.Playing:
                return "playing";
            case MediaPlayer.Event.Paused:
                return "paused";
            case MediaPlayer.Event.EndReached:
                return "ended";
            case MediaPlayer.Event.EncounteredError:
                return "error";
            default:
                return "idle";
        }
    }
    
    // MARK: - Event Handling
    
    @Override
    public void onEvent(MediaPlayer.Event event) {
        String newState = getStateFromVLC(event.type);
        
        switch (event.type) {
            case MediaPlayer.Event.MediaChanged:
                sendEvent("onLoad", createLoadData());
                sendEvent("onReady", null);
                break;
                
            case MediaPlayer.Event.Playing:
                currentState = "playing";
                sendEvent("onPlay", null);
                sendStateChangeEvent(newState);
                break;
                
            case MediaPlayer.Event.Paused:
                currentState = "paused";
                sendEvent("onPause", null);
                sendStateChangeEvent(newState);
                break;
                
            case MediaPlayer.Event.Stopped:
                currentState = "stopped";
                currentTime = 0;
                sendEvent("onStop", null);
                sendStateChangeEvent(newState);
                break;
                
            case MediaPlayer.Event.EndReached:
                currentState = "ended";
                if (loop) {
                    play();
                } else {
                    sendEvent("onEnd", null);
                }
                sendStateChangeEvent(newState);
                break;
                
            case MediaPlayer.Event.TimeChanged:
                currentTime = mediaPlayer.getTime();
                duration = mediaPlayer.getLength();
                sendEvent("onProgress", createProgressData());
                break;
                
            case MediaPlayer.Event.EncounteredError:
                currentState = "error";
                sendError("PLAYBACK_ERROR", "Erreur de lecture VLC");
                sendStateChangeEvent(newState);
                break;
        }
    }
    
    private void sendEvent(String eventName, @Nullable WritableMap data) {
        ReactContext reactContext = (ReactContext) getContext();
        reactContext.getJSModule(RCTEventEmitter.class)
                .receiveEvent(getId(), eventName, data);
    }
    
    private void sendStateChangeEvent(String newState) {
        WritableMap data = Arguments.createMap();
        data.putString("state", newState);
        data.putString("previousState", currentState);
        sendEvent("onStateChange", data);
    }
    
    private void sendError(String code, String message) {
        WritableMap error = Arguments.createMap();
        error.putString("code", code);
        error.putString("message", message);
        error.putString("domain", "VLCPlayer");
        
        WritableMap data = Arguments.createMap();
        data.putMap("error", error);
        
        sendEvent("onError", data);
    }
    
    private WritableMap createProgressData() {
        WritableMap data = Arguments.createMap();
        data.putDouble("currentTime", currentTime);
        data.putDouble("duration", duration);
        data.putDouble("position", duration > 0 ? (double) currentTime / duration : 0);
        data.putDouble("remainingTime", Math.max(0, duration - currentTime));
        return data;
    }
    
    private WritableMap createLoadData() {
        WritableMap data = Arguments.createMap();
        data.putDouble("duration", mediaPlayer.getLength());
        data.putBoolean("canPlayFast", true);
        data.putBoolean("canPlaySlow", true);
        data.putBoolean("canPause", true);
        data.putBoolean("canSeek", true);
        return data;
    }
    
    // MARK: - Fullscreen Management
    
    public void toggleFullscreen() {
        android.util.Log.d("VLCPlayerView", "toggleFullscreen() appelé, isFullscreen=" + isFullscreen);
        
        if (isFullscreen) {
            // Sortir du plein écran (ne devrait pas arriver car géré par l'Activity)
            isFullscreen = false;
            WritableMap data = Arguments.createMap();
            data.putBoolean("isFullscreen", false);
            sendEvent("onFullscreenChange", data);
            return;
        }
        
        // Entrer en plein écran en lançant l'Activity dédiée
        Activity activity = getActivity();
        if (activity == null) {
            android.util.Log.e("VLCPlayerView", "Activity est null");
            sendError("FULLSCREEN_ERROR", "Impossible d'accéder à l'Activity pour le plein écran");
            return;
        }
        
        // Obtenir l'URI du média actuel
        String mediaUri = null;
        if (source != null && source.hasKey("uri")) {
            mediaUri = source.getString("uri");
        }
        
        if (mediaUri == null) {
            sendError("FULLSCREEN_ERROR", "Aucun média chargé pour le plein écran");
            return;
        }
        
        // Obtenir le temps actuel
        long currentTime = 0;
        if (mediaPlayer != null) {
            currentTime = mediaPlayer.getTime();
        }
        
        android.util.Log.d("VLCPlayerView", "Lancement de l'Activity plein écran avec URI: " + mediaUri + " et temps: " + currentTime);
        
        // Lancer l'Activity plein écran
        try {
            android.content.Intent intent = new android.content.Intent(activity, VLCFullscreenActivity.class);
            intent.putExtra(VLCFullscreenActivity.EXTRA_MEDIA_URI, mediaUri);
            intent.putExtra(VLCFullscreenActivity.EXTRA_CURRENT_TIME, currentTime);
            activity.startActivity(intent);
            
            // Marquer comme en plein écran
            isFullscreen = true;
            WritableMap data = Arguments.createMap();
            data.putBoolean("isFullscreen", true);
            sendEvent("onFullscreenChange", data);
            
            android.util.Log.d("VLCPlayerView", "Activity plein écran lancée avec succès");
            
        } catch (Exception e) {
            android.util.Log.e("VLCPlayerView", "Erreur lors du lancement de l'Activity plein écran", e);
            sendError("FULLSCREEN_ERROR", "Impossible de lancer l'Activity plein écran: " + e.getMessage());
        }
    }
    
    // MARK: - Picture-in-Picture Management
    
    public void enterPictureInPicture() {
        android.util.Log.d("VLCPlayerView", "enterPictureInPicture() appelé");
        
        // Vérifier si PiP est supporté (API 24+)
        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.N) {
            android.util.Log.w("VLCPlayerView", "Picture-in-Picture non supporté sur cette version d'Android (API < 24)");
            sendError("PIP_NOT_SUPPORTED", "Picture-in-Picture nécessite Android 7.0 ou plus récent");
            return;
        }
        
        Activity activity = getActivity();
        if (activity == null) {
            android.util.Log.e("VLCPlayerView", "Activity est null");
            sendError("PIP_ERROR", "Impossible d'accéder à l'Activity pour le Picture-in-Picture");
            return;
        }
        
        try {
            // Créer les paramètres PiP
            PictureInPictureParams.Builder pipBuilder = new PictureInPictureParams.Builder();
            
            // Définir le ratio d'aspect (16:9 par défaut)
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
                pipBuilder.setAspectRatio(new Rational(16, 9));
            }
            
            // Entrer en mode Picture-in-Picture
            boolean success = activity.enterPictureInPictureMode(pipBuilder.build());
            
            if (success) {
                android.util.Log.d("VLCPlayerView", "Picture-in-Picture activé avec succès");
                
                // Notifier React Native
                WritableMap data = Arguments.createMap();
                data.putBoolean("isPictureInPicture", true);
                sendEvent("onPictureInPictureChange", data);
                
            } else {
                android.util.Log.w("VLCPlayerView", "Échec de l'activation du Picture-in-Picture");
                sendError("PIP_FAILED", "Impossible d'activer le Picture-in-Picture");
            }
            
        } catch (Exception e) {
            android.util.Log.e("VLCPlayerView", "Erreur lors de l'activation du Picture-in-Picture", e);
            sendError("PIP_ERROR", "Erreur Picture-in-Picture: " + e.getMessage());
        }
    }
    
    private Activity getActivity() {
        Context context = getContext();
        while (context instanceof ReactContext) {
            ReactContext reactContext = (ReactContext) context;
            if (reactContext.getCurrentActivity() != null) {
                return reactContext.getCurrentActivity();
            }
            context = reactContext.getBaseContext();
        }
        
        if (context instanceof Activity) {
            return (Activity) context;
        }
        
        return null;
    }
    
    // MARK: - Cleanup
    
    @Override
    protected void onDetachedFromWindow() {
        super.onDetachedFromWindow();
        cleanup();
    }
    
    private void cleanup() {
        if (mediaPlayer != null) {
            mediaPlayer.stop();
            mediaPlayer.detachViews();
            mediaPlayer.release();
            mediaPlayer = null;
        }
        
        if (libVLC != null) {
            libVLC.release();
            libVLC = null;
        }
    }
} 